pkg/controller/runtimecluster/garden/reconciler.go

```diff
@@ -0,0 +1,184 @@
+// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package garden
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	certv1alpha1 "github.com/gardener/cert-management/pkg/apis/cert/v1alpha1"
+	"github.com/gardener/cert-management/pkg/cert/source"
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	operatorv1alpha1 "github.com/gardener/gardener/pkg/apis/operator/v1alpha1"
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	"github.com/gardener/gardener/pkg/controllerutils"
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/gardener/gardener-extension-shoot-cert-service/pkg/controller/runtimecluster/certificate"
+)
+
+// Reconciler reconciles Gardens.
+type Reconciler struct {
+	RuntimeClientSet  kubernetes.Interface
+	ControllerOptions controller.Options
+	GardenNamespace   string
+}
+
+// Reconcile performs the main reconciliation logic.
+func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
```

Comment:
Why is this controller reconciling a `garden` object? Actually, it's supposed to reconcile an `Extension` object created by `gardener-operator`.

Reply:
The `garden` object is reconciled to react on updates on `spec.virtualCluster.dns.domains` and `spec.runtimeCluster.ingress.domains`. This means the runtime `Certificate` resource is updated accordingly to reflect these changes.

Reconciling the extension object created by gardener-operator only provides the deployment of the cert-controller-manager. To enable automatic creation of the certificate, one need to set `.spec.runtimeClusterValues.gardenerCertificates.runtimeCluster.enabled=true`.
I have added documentation for that.